io-sim/src/Control/Monad/IOSim/Internal.hs

```diff
@@ -332,38 +335,80 @@ schedule !thread@Thread{
                                           , nextTmid = succ nextTmid }
       return (SimTrace time tid tlbl (EventTimerCreated nextTmid nextVid expiry) trace)
 
+    NewRegisterDelay d k | d < 0 ->
+      {-# SCC "schedule.NewRegisterDelay" #-} do
+      !tvar <- execNewTVar nextVid
+                          (Just $ "<<timeout " ++ show (unTimeoutId nextTmid) ++ ">>")
+                          True
+      let !expiry  = d `addTime` time
+          !thread' = thread { threadControl = ThreadControl (k tvar) ctl }
+      !trace <- schedule thread' simstate { nextVid = succ nextVid }
+      return (SimTrace time tid tlbl (EventRegisterDelayCreated nextTmid nextVid expiry) $
+              SimTrace time tid tlbl (EventTimerCancelled nextTmid) $
+              trace)
+
+    NewRegisterDelay d k ->
+      {-# SCC "schedule.NewRegisterDelay" #-} do
+      !tvar <- execNewTVar nextVid
+                          (Just $ "<<timeout " ++ show (unTimeoutId nextTmid) ++ ">>")
+                          False
+      let !expiry  = d `addTime` time
+          !timers' = PSQ.insert nextTmid expiry (TimerRegisterDelay tvar) timers
+          !thread' = thread { threadControl = ThreadControl (k tvar) ctl }
+      !trace <- schedule thread' simstate { timers   = timers'
+                                          , nextVid  = succ nextVid
+                                          , nextTmid = succ nextTmid }
+      return (SimTrace time tid tlbl
+                (EventRegisterDelayCreated nextTmid nextVid expiry) trace)
+
+    NewThreadDelay d k | d < 0 ->
+      {-# SCC "schedule.NewThreadDelay" #-} do
+      let !expiry  = d `addTime` time
+          !thread' = thread { threadControl = ThreadControl k ctl }
+      !trace <- schedule thread' simstate
+      return (SimTrace time tid tlbl (EventThreadDelay nextTmid expiry) $
+              SimTrace time tid tlbl (EventTimerCancelled nextTmid) $
+              trace)
+
+    NewThreadDelay d k ->
+      {-# SCC "schedule.NewThreadDelay" #-} do
+      let !expiry  = d `addTime` time
+          !timers' = PSQ.insert nextTmid expiry (TimerThreadDelay tid) timers
+          !thread' = thread { threadControl = ThreadControl k ctl }
+      !trace <- deschedule Blocked thread' simstate { timers   = timers'
+                                                    , nextTmid = succ nextTmid }
+      return (SimTrace time tid tlbl (EventThreadDelay nextTmid expiry) trace)
+
+
     -- we do not follow `GHC.Event` behaviour here; updating a timer to the past
     -- effectively cancels it.
-    UpdateTimeout (Timeout _tvar _tvar' tmid) d k | d < 0 ->
+    UpdateTimeout (Timeout _tvar tmid) d k | d < 0 ->
       {-# SCC "schedule.UpdateTimeout" #-} do
       let !timers' = PSQ.delete tmid timers
           !thread' = thread { threadControl = ThreadControl k ctl }
-      trace <- schedule thread' simstate { timers = timers' }
+      !trace <- schedule thread' simstate { timers = timers' }
```

Comment:
Same with all of these: I'm not convinced they should be strict, but even if it should, that should be in its own patch, with a commit that explains why.

Reply:
Done